Modules/posixmodule.c

```diff
@@ -5088,384 +5088,392 @@ os__path_splitroot_impl(PyObject *module, path_t *path)
 }
 
 
-/*[clinic input]
-os._path_isdir
-
-    s: 'O'
+#define PY_IFREG  1 // Regular file
+#define PY_IFDIR  2 // Directory
+#define PY_IFLNK  4 // Symlink
+#define PY_IFMNT  8 // Mount Point (junction)
+#define PY_IFLRP 16 // Link Reparse Point (name-surrogate, symlink, junction)
+#define PY_IFRRP 32 // Regular Reparse Point
 
-Return true if the pathname refers to an existing directory.
-
-[clinic start generated code]*/
+static inline BOOL
+_testReparseTag(DWORD reparseTag, int testedType)
+{
+    switch (testedType) {
+    case PY_IFLNK:
+        return reparseTag == IO_REPARSE_TAG_SYMLINK;
+    case PY_IFMNT:
+        return reparseTag == IO_REPARSE_TAG_MOUNT_POINT;
+    case PY_IFLRP:
+        return IsReparseTagNameSurrogate(reparseTag);
+    case PY_IFRRP:
+        return reparseTag && !IsReparseTagNameSurrogate(reparseTag);
+    }
+    return FALSE
+}
 
-static PyObject *
-os__path_isdir_impl(PyObject *module, PyObject *s)
-/*[clinic end generated code: output=9d87ab3c8b8a4e61 input=c17f7ef21d22d64e]*/
+static BOOL
+_testFileTypeByHandle(HANDLE hfile, int testedType, BOOL diskOnly)
 {
-    HANDLE hfile;
-    BOOL close_file = TRUE;
-    FILE_BASIC_INFO info;
-    path_t _path = PATH_T_INITIALIZE("isdir", "s", 0, 1);
-    int result;
-    BOOL slow_path = TRUE;
-    FILE_STAT_BASIC_INFORMATION statInfo;
+    assert(testedType == PY_IFREG || testedType == PY_IFDIR ||
+           testedType == PY_IFLNK || testedType == PY_IFMNT ||
+           testedType == PY_IFLRP || testedType == PY_IFRRP);
 
-    if (!path_converter(s, &_path)) {
-        path_cleanup(&_path);
-        if (PyErr_ExceptionMatches(PyExc_ValueError)) {
-            PyErr_Clear();
-            Py_RETURN_FALSE;
-        }
-        return NULL;
+    DWORD fileDevType = GetFileType(hfile);
+    if ((fileDevType == FILE_TYPE_UNKNOWN && GetLastError()) ||
+        (fileDevType == FILE_TYPE_CHAR) ||
+        (diskOnly && fileDevType != FILE_TYPE_DISK))
+    {
+        return FALSE;
     }
```

Comment:
`fileDevType` won't be needed later if `attributes` is checked instead. Also, bad handles and kernel file objects opened for character devices will fail in the `GetFileInformationByHandleEx()` calls. So this check is only needed for the `diskOnly` case.

```suggestion
    if (diskOnly && GetFileType(hfile) != FILE_TYPE_DISK) {
        return FALSE;
    }
```